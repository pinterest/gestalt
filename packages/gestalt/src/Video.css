.video {
  composes: block from "./Layout.css";
  composes: absolute from "./Layout.css";
  composes: xsCol12 from "./Column.css";
  height: 100%;
}

.player {
  composes: relative from "./Layout.css";
  composes: xsCol12 from "./Column.css";
  background-color: #000;
}

.playhead {
  composes: flex from "./Layout.css";
  composes: itemsCenter from "./Layout.css";
  composes: xsCol12 from "./Column.css";
  composes: pointer from "./Cursor.css";
  height: 100%;
}

.controls {
  composes: absolute from "./Layout.css";
  composes: flex from "./Layout.css";
  composes: itemsCenter from "./Layout.css";
  composes: bottom0 from "./Layout.css";
  composes: left0 from "./Layout.css";
  composes: right0 from "./Layout.css";
  composes: paddingX2 from "./boxWhitespace.css";
  composes: paddingY2 from "./boxWhitespace.css";
  background-image: linear-gradient(to bottom, rgba(181, 181, 181, 0), rgba(0, 0, 0, .2) 33%, rgba(0, 0, 0, .5) 66%, rgba(0, 0, 0, .7));
}
