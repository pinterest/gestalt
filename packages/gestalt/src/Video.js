// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import VideoControls from './VideoControls.js';
import styles from './Video.css';

type Source =
  | string
  | Array<{| type: 'video/m3u8' | 'video/mp4' | 'video/ogg', src: string |}>;

type Props = {|
  accessibilityMaximizeLabel: string,
  accessibilityMinimizeLabel: string,
  accessibilityMuteLabel: string,
  accessibilityPauseLabel: string,
  accessibilityPlayLabel: string,
  accessibilityUnmuteLabel: string,
  aspectRatio: number,
  captions: string,
  controls?: boolean,
  loop?: boolean,
  onDurationChange?: ({
    event: SyntheticEvent<HTMLVideoElement>,
    duration: number,
  }) => void,
  onEnded?: ({ event: SyntheticEvent<HTMLVideoElement> }) => void,
  onFullscreenChange?: ({ event: Event, fullscreen: boolean }) => void,
  onLoadedChange?: ({
    event: SyntheticEvent<HTMLVideoElement>,
    loaded: number,
  }) => void,
  onPlay?: ({ event: SyntheticEvent<HTMLDivElement> }) => void,
  onPause?: ({ event: SyntheticEvent<HTMLDivElement> }) => void,
  onReady?: ({ event: SyntheticEvent<HTMLVideoElement> }) => void,
  onSeek?: ({ event: SyntheticEvent<HTMLVideoElement> }) => void,
  onTimeChange?: ({
    event: SyntheticEvent<HTMLVideoElement>,
    time: number,
  }) => void,
  onVolumeChange?: ({
    event: SyntheticEvent<HTMLDivElement>,
    volume: number,
  }) => void,
  playbackRate: number,
  playing: boolean,
  playsInline?: boolean,
  poster?: string,
  preload: 'auto' | 'metadata' | 'none',
  src: Source,
  volume: number,
|};

type State = {|
  currentTime: number,
  duration: number,
  fullscreen: boolean,
|};

// For more information on fullscreen and vendor prefixes see
// https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API

const requestFullscreen = (element: HTMLElement) => {
  if (element.requestFullscreen) {
    element.requestFullscreen();
    // $FlowIssue - vendor prefix missing from Flow
  } else if (element.webkitRequestFullscreen) {
    // $FlowIssue - vendor prefix missing from Flow
    element.webkitRequestFullscreen();
    // $FlowIssue - vendor prefix missing from Flow
  } else if (element.mozRequestFullScreen) {
    // $FlowIssue - vendor prefix missing from Flow
    element.mozRequestFullScreen();
    // $FlowIssue - vendor prefix missing from Flow
  } else if (element.msRequestFullscreen) {
    // $FlowIssue - vendor prefix missing from Flow
    element.msRequestFullscreen();
  }
};

const exitFullscreen = () => {
  // $FlowIssue - vendor prefix missing from Flow
  if (document.exitFullscreen) {
    // $FlowIssue - vendor prefix missing from Flow
    document.exitFullscreen();
    // $FlowIssue - vendor prefix missing from Flow
  } else if (document.webkitExitFullscreen) {
    // $FlowIssue - vendor prefix missing from Flow
    document.webkitExitFullscreen();
    // $FlowIssue - vendor prefix missing from Flow
  } else if (document.mozCancelFullScreen) {
    // $FlowIssue - vendor prefix missing from Flow
    document.mozCancelFullScreen();
    // $FlowIssue - vendor prefix missing from Flow
  } else if (document.msExitFullscreen) {
    // $FlowIssue - vendor prefix missing from Flow
    document.msExitFullscreen();
  }
};

// Normally document.fullscreen suffices here as a flag, but IE11 does not
// have a vendor specific version so we must instead use the actual element
const isFullscreen = () =>
  // $FlowIssue - vendor prefix missing from Flow
  document.fullscreenElement ||
  // $FlowIssue - vendor prefix missing from Flow
  document.webkitFullscreenElement ||
  // $FlowIssue - vendor prefix missing from Flow
  document.mozFullScreenElement ||
  // $FlowIssue - vendor prefix missing from Flow
  document.msFullscreenElement;

const addFullscreenEventListener = (listener: EventListener) => {
  document.addEventListener('fullscreenchange', listener);
  document.addEventListener('webkitfullscreenchange', listener);
  document.addEventListener('mozfullscreenchange', listener);
  document.addEventListener('MSFullscreenChange', listener);
};

const removeFullscreenEventListener = (listener: EventListener) => {
  document.removeEventListener('fullscreenchange', listener);
  document.removeEventListener('webkitfullscreenchange', listener);
  document.removeEventListener('mozfullscreenchange', listener);
  document.removeEventListener('MSFullscreenChange', listener);
};

const isNewSource = (oldSource: Source, newSource: Source): boolean => {
  if (typeof oldSource !== typeof newSource) {
    // If the source type changed from string to Array
    // or vice versa, we have a new source
    return true;
  }
  if (Array.isArray(newSource)) {
    if (oldSource.length !== newSource.length) {
      // If the sources are both an Array, and the lengths
      // do not match we evaluate as a new source
      return true;
    }
    // If the sources are both an Array and the same length,
    // verify every element stayed the same
    return newSource.some(
      (source, index) =>
        !Array.isArray(oldSource) ||
        source.type !== oldSource[index].type ||
        source.src !== oldSource[index].src
    );
  }
  // If the sources are both a string, simply compare
  // the new with the old
  return newSource !== oldSource;
};

export default class Video extends React.PureComponent<Props, State> {
  static propTypes = {
    accessibilityMaximizeLabel: PropTypes.string,
    accessibilityMinimizeLabel: PropTypes.string,
    accessibilityMuteLabel: PropTypes.string,
    accessibilityPauseLabel: PropTypes.string,
    accessibilityPlayLabel: PropTypes.string,
    accessibilityUnmuteLabel: PropTypes.string,
    aspectRatio: PropTypes.number.isRequired,
    captions: PropTypes.string.isRequired,
    controls: PropTypes.bool,
    loop: PropTypes.bool,
    onDurationChange: PropTypes.func,
    onEnded: PropTypes.func,
    onFullscreenChange: PropTypes.func,
    onLoadedChange: PropTypes.func,
    onPlay: PropTypes.func,
    onPause: PropTypes.func,
    onReady: PropTypes.func,
    onSeek: PropTypes.func,
    onTimeChange: PropTypes.func,
    onVolumeChange: PropTypes.func,
    playbackRate: PropTypes.number,
    playing: PropTypes.bool,
    playsInline: PropTypes.bool,
    poster: PropTypes.string,
    preload: PropTypes.oneOf(['auto', 'metadata', 'none']),
    src: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.arrayOf(
        PropTypes.shape({
          type: PropTypes.oneOf(['video/m3u8', 'video/mp4', 'video/ogg'])
            .isRequired,
          src: PropTypes.string.isRequired,
        })
      ),
    ]).isRequired,
    volume: PropTypes.number,
  };

  static defaultProps = {
    playbackRate: 1,
    playing: false,
    preload: 'auto',
    volume: 1,
  };

  state = {
    currentTime: 0,
    duration: 0,
    fullscreen: false,
  };

  /**
   * React lifecycle hooks pertinent to Video
   */

  componentDidMount() {
    const { playbackRate, playing, volume } = this.props;
    // Set up event listeners to catch backdoors in fullscreen
    // changes such as using the ESC key to exit
    if (typeof document !== 'undefined') {
      addFullscreenEventListener(this.handleFullscreenChange);
    }
    // Load the video to hydrate the DOM after a server render
    this.load();
    // Set the initial volume
    this.setVolume(volume);
    // Set the initial playback rate
    this.setPlaybackRate(playbackRate);
    // Simulate an autoplay effect if the component
    if (playing) {
      this.play();
    }
  }

  componentDidUpdate(prevProps: Props) {
    // If the video source changed, reload the video
    if (isNewSource(prevProps.src, this.props.src)) {
      this.load();
    }
    // If the volume changed, set the new volume
    if (prevProps.volume !== this.props.volume) {
      this.setVolume(this.props.volume);
    }
    // If the playback rate changed, set the new rate
    if (prevProps.playbackRate !== this.props.playbackRate) {
      this.setPlaybackRate(this.props.playbackRate);
    }
    // If the playback changed, play or pause the video
    if (prevProps.playing !== this.props.playing) {
      if (this.props.playing) {
        this.play();
      } else {
        this.pause();
      }
    }
  }

  componentWillUnmount() {
    removeFullscreenEventListener(this.handleFullscreenChange);
  }

  /**
   * DOM reference housekeeping that is needed for functionality
   */

  // The player element encapsulates the actual video DOM
  // element as well as the controls to bring both fullscreen
  setPlayerRef = (ref: ?HTMLDivElement) => {
    this.player = ref;
  };

  // The actual reference to the video HTML DOM element
  setVideoRef = (ref: ?HTMLVideoElement) => {
    this.video = ref;
  };

  /**
   * Functions that directly interact with the HTML video element
   */

  // Set the video to the desired playback rate: 1 (normal)
  setPlaybackRate = (playbackRate: number) => {
    if (this.video) {
      this.video.playbackRate = playbackRate;
    }
  };

  // Set the video to the desired volume: 0 (muted) -> 1 (max)
  setVolume = (volume: number) => {
    if (this.video) {
      this.video.volume = volume;
    }
  };

  // Change the video source and re-load the video
  load = () => {
    if (this.video) {
      this.video.load();
    }
  };

  // Pause the video
  pause = () => {
    if (this.video) {
      this.video.pause();
    }
  };

  // Play the video
  play = () => {
    if (this.video) {
      this.video.play();
    }
  };

  // Seek the video to the desired time
  seek = (time: number) => {
    if (this.video) {
      this.video.currentTime = time;
    }
  };

  // Enter/exit fullscreen video player mode
  toggleFullscreen = () => {
    if (isFullscreen()) {
      exitFullscreen();
    } else if (this.player) {
      requestFullscreen(this.player);
    }
  };

  /**
   * Handlers for various media events on the video
   */

  // Sent when enough data is available that the media can be played
  handleCanPlay = (event: SyntheticEvent<HTMLVideoElement>) => {
    const { onReady } = this.props;

    if (onReady) {
      onReady({ event });
    }
  };

  // The metadata has loaded or changed, indicating a change in
  // duration of the media
  handleDurationChange = (event: SyntheticEvent<HTMLVideoElement>) => {
    const { onDurationChange } = this.props;
    const duration = (this.video && this.video.duration) || 0;
    this.setState({ duration });

    if (onDurationChange) {
      onDurationChange({ event, duration });
    }
  };

  // Sent when playback completes.
  handleEnded = (event: SyntheticEvent<HTMLVideoElement>) => {
    const { onEnded } = this.props;

    if (onEnded) {
      onEnded({ event });
    }
  };

  // Sent when the video is switched to/out-of fullscreen mode
  handleFullscreenChange = (event: Event) => {
    const { onFullscreenChange } = this.props;
    const fullscreen = !!isFullscreen();
    this.setState({ fullscreen });

    if (onFullscreenChange) {
      onFullscreenChange({ event, fullscreen });
    }
  };

  // Sent when playback of the media starts after having been paused.
  handlePlay = (event: SyntheticEvent<HTMLDivElement>) => {
    const { onPlay } = this.props;

    if (onPlay) {
      onPlay({ event });
    }
  };

  // Sent when playback is paused.
  handlePause = (event: SyntheticEvent<HTMLDivElement>) => {
    const { onPause } = this.props;

    if (onPause) {
      onPause({ event });
    }
  };

  // Sent periodically to inform interested parties of progress downloading the media
  handleProgress = (event: SyntheticEvent<HTMLVideoElement>) => {
    const { onLoadedChange } = this.props;
    const { buffered } = this.video || {};
    const loaded =
      buffered && buffered.length > 0 ? buffered.end(buffered.length - 1) : 0;

    if (onLoadedChange) {
      onLoadedChange({ event, loaded });
    }
  };

  // Sent when a seek operation completes.
  handleSeek = (event: SyntheticEvent<HTMLVideoElement>) => {
    const { onSeek } = this.props;

    if (onSeek) {
      onSeek({ event });
    }
  };

  // The time indicated by the element's currentTime attribute has changed
  handleTimeUpdate = (event: SyntheticEvent<HTMLVideoElement>) => {
    const { onTimeChange } = this.props;
    const currentTime = (this.video && this.video.currentTime) || 0;
    this.setState({ currentTime });

    if (onTimeChange) {
      onTimeChange({ event, time: currentTime });
    }
  };

  // Sent when the audio volume changes
  handleVolumeChange = (event: SyntheticEvent<HTMLDivElement>) => {
    const { onVolumeChange } = this.props;
    const muted = (this.video && this.video.muted) || false;

    if (onVolumeChange) {
      onVolumeChange({ event, volume: muted ? 1 : 0 });
    }
  };

  video: ?HTMLVideoElement;

  player: ?HTMLDivElement;

  render() {
    const {
      aspectRatio,
      captions,
      loop,
      playing,
      playsInline,
      poster,
      preload,
      src,
      volume,
    } = this.props;
    const { currentTime, duration, fullscreen } = this.state;

    const paddingBottom = (fullscreen && '0') || `${(1 / aspectRatio) * 100}%`;

    return (
      <div
        ref={this.setPlayerRef}
        className={styles.player}
        style={{ paddingBottom, height: fullscreen ? '100%' : 0 }}
      >
        <video
          autoPlay={playing}
          loop={loop}
          muted={volume === 0}
          playsInline={playsInline}
          poster={poster}
          preload={preload}
          src={typeof src === 'string' ? src : undefined}
          ref={this.setVideoRef}
          className={styles.video}
          onCanPlay={this.handleCanPlay}
          onDurationChange={this.handleDurationChange}
          onEnded={this.handleEnded}
          onSeeked={this.handleSeek}
          onTimeUpdate={this.handleTimeUpdate}
          onProgress={this.handleProgress}
        >
          {Array.isArray(src) &&
            src.map(source => (
              <source key={source.src} src={source.src} type={source.type} />
            ))}
          <track kind="captions" src={captions} />
        </video>
        {/* Need to use full path for these props so Flow can infer correct subtype */}
        {this.props.controls && (
          <VideoControls
            accessibilityMaximizeLabel={this.props.accessibilityMaximizeLabel}
            accessibilityMinimizeLabel={this.props.accessibilityMinimizeLabel}
            accessibilityMuteLabel={this.props.accessibilityMuteLabel}
            accessibilityPauseLabel={this.props.accessibilityPauseLabel}
            accessibilityPlayLabel={this.props.accessibilityPlayLabel}
            accessibilityUnmuteLabel={this.props.accessibilityUnmuteLabel}
            currentTime={currentTime}
            duration={duration}
            fullscreen={fullscreen}
            onPlay={this.handlePlay}
            onPause={this.handlePause}
            onFullscreenChange={this.toggleFullscreen}
            onVolumeChange={this.handleVolumeChange}
            playing={playing}
            seek={this.seek}
            volume={volume}
          />
        )}
      </div>
    );
  }
}
