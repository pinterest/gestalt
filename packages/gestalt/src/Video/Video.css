.video {
  composes: block from "../Layout.css";
  composes: absolute from "../Layout.css";
  composes: xsCol12 from "../Column.css";
  height: 100%;
}

.player {
  composes: relative from "../Layout.css";
  composes: xsCol12 from "../Column.css";
  background-color: #000;
}

.playhead {
  composes: flex from "../Layout.css";
  composes: itemsCenter from "../Layout.css";
  composes: xsCol12 from "../Column.css";
  composes: pointer from "../Cursor.css";
  height: 100%;
}
