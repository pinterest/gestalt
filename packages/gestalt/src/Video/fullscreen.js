// @flow

// For more information on fullscreen and vendor prefixes see
// https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API

export const request = (element: HTMLElement) => {
  if (element.requestFullscreen) {
    element.requestFullscreen();
  } else if (element.webkitRequestFullscreen) {
    // $FlowIssue - missing from Flow
    element.webkitRequestFullscreen();
  } else if (element.mozRequestFullScreen) {
    // $FlowIssue - missing from Flow
    element.mozRequestFullScreen();
  } else if (element.msRequestFullscreen) {
    // $FlowIssue - missing from Flow
    element.msRequestFullscreen();
  }
};

export const exit = () => {
  if (document.exitFullscreen) {
    // $FlowIssue - missing from Flow
    document.exitFullscreen();
  } else if (document.webkitExitFullscreen) {
    // $FlowIssue - missing from Flow
    document.webkitExitFullscreen();
  } else if (document.mozCancelFullScreen) {
    // $FlowIssue - missing from Flow
    document.mozCancelFullScreen();
  } else if (document.msExitFullscreen) {
    // $FlowIssue - missing from Flow
    document.msExitFullscreen();
  }
};

export const enabled = () =>
  document.fullscreenEnabled ||
  document.webkitFullscreenEnabled ||
  document.mozFullScreenEnabled ||
  // $FlowIssue - missing from Flow
  document.msFullscreenEnabled;

// Normally document.fullscreen suffices here as a flag, but IE11 does not
// have a vendor specific version so we must instead use the actual element
export const isFullscreen = () =>
  document.fullscreenElement ||
  document.webkitFullscreenElement ||
  document.mozFullScreenElement ||
  // $FlowIssue - missing from Flow
  document.msFullscreenElement;

export const addEventListener = (handler: Function) => {
  document.addEventListener('fullscreenchange', handler);
  document.addEventListener('webkitfullscreenchange', handler);
  document.addEventListener('mozfullscreenchange', handler);
  document.addEventListener('MSFullscreenChange', handler);
};

export const removeEventListener = (handler: Function) => {
  document.removeEventListener('fullscreenchange', handler);
  document.removeEventListener('webkitfullscreenchange', handler);
  document.removeEventListener('mozfullscreenchange', handler);
  document.removeEventListener('MSFullscreenChange', handler);
};

export default undefined;
