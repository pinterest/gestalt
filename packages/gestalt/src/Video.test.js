// @flow
import React from 'react';
import { create } from 'react-test-renderer';
import { shallow } from 'enzyme';
import Video from './Video.js';

const A11Y_LABELS = Object.freeze({
  accessibilityMaximizeLabel: 'Maximize',
  accessibilityMinimizeLabel: 'Minimize',
  accessibilityMuteLabel: 'Mute',
  accessibilityPauseLabel: 'Pause',
  accessibilityPlayLabel: 'Play',
  accessibilityUnmuteLabel: 'Unmute',
});

test('Video with source', () => {
  const tree = create(
    <Video
      {...A11Y_LABELS}
      aspectRatio={1}
      captions="https://media.w3.org/2010/05/sintel/captions.vtt"
      src="https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Video with multiple sources', () => {
  const tree = create(
    <Video
      {...A11Y_LABELS}
      aspectRatio={1}
      captions="https://media.w3.org/2010/05/sintel/captions.vtt"
      src={[
        {
          type: 'video/mp4',
          src: 'https://archive.org/download/ElephantsDream/ed_1024_512kb.mp4',
        },
        {
          type: 'video/ogg',
          src: 'https://archive.org/download/ElephantsDream/ed_hd.ogv',
        },
      ]}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Video with media attributes', () => {
  const tree = create(
    <Video
      {...A11Y_LABELS}
      aspectRatio={1}
      captions="https://media.w3.org/2010/05/sintel/captions.vtt"
      loop
      volume={0}
      preload="metadata"
      src="https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Video with callbacks', () => {
  const tree = create(
    <Video
      {...A11Y_LABELS}
      aspectRatio={1}
      captions="https://media.w3.org/2010/05/sintel/captions.vtt"
      onDurationChange={() => {}}
      onEnded={() => {}}
      onFullscreenChange={() => {}}
      onLoadedChange={() => {}}
      onPause={() => {}}
      onPlay={() => {}}
      onReady={() => {}}
      onSeek={() => {}}
      onTimeChange={() => {}}
      onVolumeChange={() => {}}
      src="https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('Video loading', () => {
  it('Does not load when string src does not change', () => {
    const wrapper = shallow(
      <Video
        {...A11Y_LABELS}
        aspectRatio={1}
        captions="https://media.w3.org/2010/05/sintel/captions.vtt"
        src="https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
      />
    );
    const spy = jest.spyOn(wrapper.instance(), 'load');
    wrapper.setProps({ volume: 0 });
    expect(spy).not.toHaveBeenCalled();
  });

  it('Does not load when array src does not change', () => {
    const wrapper = shallow(
      <Video
        {...A11Y_LABELS}
        aspectRatio={1}
        captions="https://media.w3.org/2010/05/sintel/captions.vtt"
        src={[
          {
            type: 'video/mp4',
            src: 'http://media.w3.org/2010/05/bunny/movie.mp4',
          },
        ]}
      />
    );
    const spy = jest.spyOn(wrapper.instance(), 'load');
    wrapper.setProps({ volume: 0 });
    expect(spy).not.toHaveBeenCalled();
  });

  it('Loads when string src changes to new string src', () => {
    const wrapper = shallow(
      <Video
        {...A11Y_LABELS}
        aspectRatio={1}
        captions="https://media.w3.org/2010/05/sintel/captions.vtt"
        src="https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
      />
    );
    const spy = jest.spyOn(wrapper.instance(), 'load');
    wrapper.setProps({ src: 'http://media.w3.org/2010/05/bunny/movie.mp4' });
    expect(spy).toHaveBeenCalled();
  });

  it('Loads when string src changes to new array src', () => {
    const wrapper = shallow(
      <Video
        {...A11Y_LABELS}
        aspectRatio={1}
        captions="https://media.w3.org/2010/05/sintel/captions.vtt"
        src="https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
      />
    );
    const spy = jest.spyOn(wrapper.instance(), 'load');
    wrapper.setProps({
      src: [
        {
          type: 'video/mp4',
          src: 'http://media.w3.org/2010/05/bunny/movie.mp4',
        },
      ],
    });
    expect(spy).toHaveBeenCalled();
  });

  it('Loads when array src changes to new string src', () => {
    const wrapper = shallow(
      <Video
        {...A11Y_LABELS}
        aspectRatio={1}
        captions="https://media.w3.org/2010/05/sintel/captions.vtt"
        src={[
          {
            type: 'video/mp4',
            src: 'http://media.w3.org/2010/05/bunny/movie.mp4',
          },
        ]}
      />
    );
    const spy = jest.spyOn(wrapper.instance(), 'load');
    wrapper.setProps({
      src: 'https://media.w3.org/2010/05/sintel/trailer_hd.mp4',
    });
    expect(spy).toHaveBeenCalled();
  });

  it('Loads when array src changes to new array src', () => {
    const wrapper = shallow(
      <Video
        {...A11Y_LABELS}
        aspectRatio={1}
        captions="https://media.w3.org/2010/05/sintel/captions.vtt"
        src={[
          {
            type: 'video/mp4',
            src: 'https://media.w3.org/2010/05/sintel/trailer_hd.mp4',
          },
        ]}
      />
    );
    const spy = jest.spyOn(wrapper.instance(), 'load');
    wrapper.setProps({
      src: [
        {
          type: 'video/mp4',
          src: 'http://media.w3.org/2010/05/bunny/movie.mp4',
        },
      ],
    });
    expect(spy).toHaveBeenCalled();
  });

  it('Loads when array src changes to new length array src', () => {
    const wrapper = shallow(
      <Video
        {...A11Y_LABELS}
        aspectRatio={1}
        captions="https://media.w3.org/2010/05/sintel/captions.vtt"
        src={[
          {
            type: 'video/mp4',
            src: 'https://media.w3.org/2010/05/sintel/trailer_hd.mp4',
          },
        ]}
      />
    );
    const spy = jest.spyOn(wrapper.instance(), 'load');
    wrapper.setProps({
      src: [
        {
          type: 'video/mp4',
          src: 'https://archive.org/download/ElephantsDream/ed_1024_512kb.mp4',
        },
        {
          type: 'video/ogg',
          src: 'https://archive.org/download/ElephantsDream/ed_hd.ogv',
        },
      ],
    });
    expect(spy).toHaveBeenCalled();
  });
});
