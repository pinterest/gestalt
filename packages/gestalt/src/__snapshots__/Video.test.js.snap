// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Video with callbacks 1`] = `
<div
  className="player"
  style={
    Object {
      "height": 0,
      "paddingBottom": "100%",
    }
  }
>
  <video
    autoPlay={false}
    className="video"
    muted={false}
    onCanPlay={[Function]}
    onDurationChange={[Function]}
    onEnded={[Function]}
    onProgress={[Function]}
    onSeeked={[Function]}
    onTimeUpdate={[Function]}
    preload="auto"
    src="https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
  >
    <track
      kind="captions"
      src="https://media.w3.org/2010/05/sintel/captions.vtt"
    />
  </video>
</div>
`;

exports[`Video with children 1`] = `
<div
  className="player"
  style={
    Object {
      "height": 0,
      "paddingBottom": "100%",
    }
  }
>
  <video
    autoPlay={false}
    className="video"
    muted={false}
    onCanPlay={[Function]}
    onDurationChange={[Function]}
    onEnded={[Function]}
    onProgress={[Function]}
    onSeeked={[Function]}
    onTimeUpdate={[Function]}
    preload="auto"
    src="https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
  >
    <track
      kind="captions"
      src="https://media.w3.org/2010/05/sintel/captions.vtt"
    />
  </video>
  <div
    className="absolute bottom0 box left0 overflowHidden right0 top0"
  >
    <div>
      overlay
    </div>
  </div>
  <div
    className="controls"
  >
    <div
      className="box paddingX2 paddingY2"
    >
      <div
        className="touchable pointer square"
        onClick={[Function]}
        onKeyPress={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="button"
        tabIndex="0"
      >
        <svg
          aria-hidden={null}
          aria-label="Play"
          className="icon white iconBlock"
          height={20}
          role="img"
          viewBox="0 0 24 24"
          width={20}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </div>
    <div
      className="box paddingX2 paddingY2"
      style={
        Object {
          "width": 50,
        }
      }
    >
      <div
        className="Text fontSize1 white alignRight breakWord fontStyleNormal fontWeightNormal"
      >
        00:00
      </div>
    </div>
    <div
      className="box flexGrow paddingX2 paddingY2"
    >
      <div
        className="box flexGrow itemsCenter relative xsDisplayFlex"
        style={
          Object {
            "height": 16,
          }
        }
      >
        <div
          aria-valuemax={0}
          aria-valuemin="0"
          aria-valuenow={0}
          className="playhead"
          onClick={[Function]}
          onKeyPress={[Function]}
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          onMouseUp={[Function]}
          role="progressbar"
          tabIndex="-1"
        >
          <div
            className="absolute box left0 lightGrayBg right0 rounded"
            style={
              Object {
                "height": 4,
              }
            }
          >
            <div
              className="box rounded whiteBg"
              style={
                Object {
                  "height": "100%",
                  "width": "NaN%",
                }
              }
            />
          </div>
          <div
            className="absolute box rounded"
            style={
              Object {
                "height": 4,
                "left": "NaN%",
              }
            }
          >
            <div
              className="box circle marginLeftN2 whiteBg"
              style={
                Object {
                  "height": 16,
                  "marginTop": -6,
                  "width": 16,
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="box paddingX2 paddingY2"
      style={
        Object {
          "width": 50,
        }
      }
    >
      <div
        className="Text fontSize1 white alignRight breakWord fontStyleNormal fontWeightNormal"
      >
        00:00
      </div>
    </div>
    <div
      className="box paddingX2 paddingY2"
    >
      <div
        className="touchable pointer square"
        onClick={[Function]}
        onKeyPress={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        role="button"
        tabIndex="0"
      >
        <svg
          aria-hidden={null}
          aria-label="Mute"
          className="icon white iconBlock"
          height={20}
          role="img"
          viewBox="0 0 24 24"
          width={20}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`Video with media attributes 1`] = `
<div
  className="player"
  style={
    Object {
      "height": 0,
      "paddingBottom": "100%",
    }
  }
>
  <video
    autoPlay={false}
    className="video"
    loop={true}
    muted={true}
    onCanPlay={[Function]}
    onDurationChange={[Function]}
    onEnded={[Function]}
    onProgress={[Function]}
    onSeeked={[Function]}
    onTimeUpdate={[Function]}
    preload="metadata"
    src="https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
  >
    <track
      kind="captions"
      src="https://media.w3.org/2010/05/sintel/captions.vtt"
    />
  </video>
</div>
`;

exports[`Video with multiple sources 1`] = `
<div
  className="player"
  style={
    Object {
      "height": 0,
      "paddingBottom": "100%",
    }
  }
>
  <video
    autoPlay={false}
    className="video"
    muted={false}
    onCanPlay={[Function]}
    onDurationChange={[Function]}
    onEnded={[Function]}
    onProgress={[Function]}
    onSeeked={[Function]}
    onTimeUpdate={[Function]}
    preload="auto"
  >
    <source
      src="https://archive.org/download/ElephantsDream/ed_1024_512kb.mp4"
      type="video/mp4"
    />
    <source
      src="https://archive.org/download/ElephantsDream/ed_hd.ogv"
      type="video/ogg"
    />
    <track
      kind="captions"
      src="https://media.w3.org/2010/05/sintel/captions.vtt"
    />
  </video>
</div>
`;

exports[`Video with source 1`] = `
<div
  className="player"
  style={
    Object {
      "height": 0,
      "paddingBottom": "100%",
    }
  }
>
  <video
    autoPlay={false}
    className="video"
    muted={false}
    onCanPlay={[Function]}
    onDurationChange={[Function]}
    onEnded={[Function]}
    onProgress={[Function]}
    onSeeked={[Function]}
    onTimeUpdate={[Function]}
    preload="auto"
    src="https://media.w3.org/2010/05/sintel/trailer_hd.mp4"
  >
    <track
      kind="captions"
      src="https://media.w3.org/2010/05/sintel/captions.vtt"
    />
  </video>
</div>
`;
